### ISSUE TITLE
If a package is CamelCase, immutables will treat it as an object

### ISSUE THREAD

**Issue by ddxd** [immutables/immutables#401]
We have met the following weird issue.

in the source code, we have an import 
  import com.TaResearch.object.requests.SomeData

in the generated code, it becomes,

import com.TaResearch;
import object.requests.SomeData;

So I cannot compile the generated code successfully.

Is there an easy fix or a workaround?

**Comment by ddxd** [comment]
if I change TaResearch to taresearch then the import is perfectly fine again.

**Comment by ldriscoll** [comment]
I believe it is convention that packages should always be lower case, to avoid conflicts with classes. https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

**Comment by ddxd** [comment]
We have a legacy library that we import as jar and we cannot change the package path. I know it is stupid but it is a real case. I guess that some parsing method uses camelcase as the indicator of the end of the path and it is very robust for the wildness of people's thinking.

**Comment by elucash** [comment]
We probably should fix it, leaving the imports intact. We already have cases when we disable import post-processing when "weird" identifiers encounter. While it's not hard to fix, we are currently extremely busy with some other functionality, so the release containing the fix may be available in a week or two.
As a workaround I might suggest to use maven shade plugin (or comparable solution for other build system) to build jar with rewritten package names.

```xml
<relocation>
  <pattern>com.TaResearch</pattern>
  <shadedPattern>com.taresearch.repack</shadedPattern>
</relocation>
```

**Comment by ddxd** [comment]
Thanks. We will wait for milestone 2.2.11.

**Comment by tinkerware** [comment]
I've just run into this while generating immutable value classes that make use of the [HdrHistogram][1] library. The package name this library uses is `org.HdrHistogram`; annoyingly non-standard but a real thing that will not change at this point.

[1]: https://github.com/HdrHistogram/HdrHistogram

**Comment by tinkerware** [comment]
Also, while Maven Shade renaming works for command-line builds, it leaves at least IntelliJ unable to resolve the renamed classes. I'm not sure about other IDE's.

**Comment by elucash** [comment]
When you run maven shade plugin you create a separate jar, with repackaged information, then you meed to add this jar (or maven module) as dependency instead of original dependency jar, if done correctly IDE/compiler will perfectly see repackaged classes.
While I'm not insisting that repackaging with shade is optimal solution, and we will try to fix this in Immutables.

**Comment by ddxd** [comment]
we end with a workaround found by our smart engineer. add a
//-no-import-rewrite


*at the beginning of the file. And use *@Value.Style(headerComments = true,....)


That solves our problem.

**Comment by elucash** [comment]
wow, very smart workaround! While aware of this functionality, you've used in quite creative way!

**Comment by elucash** [comment]
Fix for this was added, but there might be still cases where something similar might show up, for those cases mentioned workaround is the best bet.
released in 2.2.11 Thank you!
